File: frontend/src/types/graph.ts
export interface GraphNode {
  id: string;
  label: string;
  node_type: string; // 'entity' | 'fact' | 'hypothesis' | 'evidence'
  sub_type?: string | null;
  attributes: Record<string, unknown>;
  source_ids: string[];
  confidence?: number;
}

export interface GraphEdge {
  id: string;
  source: string;
  target: string;
  relationship_type: 'supports' | 'attacks' | 'derives_from' | 'contradicts' | 'associated_with';
  label?: string | null;
  confidence: number;
  source_ids: string[];
  attributes: Record<string, unknown>;
}

export interface EvidenceGraph {
  case_id: string;
  node_count: number;
  edge_count: number;
  nodes: GraphNode[];
  edges: GraphEdge[];
  dedup_clusters: Record<string, string[]>;
}

File: backend/app/schemas/graph.py
from typing import Any, Dict, List, Optional

from pydantic import BaseModel, ConfigDict


class GraphNode(BaseModel):
    id: str
    label: str
    node_type: str  # "entity", "fact", "hypothesis", "evidence"
    sub_type: Optional[str] = None  # e.g. "PERSON", "PHONE_NUMBER", "OBSERVATION"
    attributes: Dict[str, Any] = {}
    source_ids: List[str] = []
    cluster_id: Optional[str] = None

    model_config = ConfigDict(from_attributes=True)


class GraphEdge(BaseModel):
    id: str
    source: str
    target: str
    relationship_type: (
        str  # "supports", "attacks", "derives_from", "contradicts", "associated_with"
    )
    label: Optional[str] = None
    confidence: float = 1.0
    source_ids: List[str] = []
    attributes: Dict[str, Any] = {}

    model_config = ConfigDict(from_attributes=True)


class EvidenceGraphResponse(BaseModel):
    case_id: str
    node_count: int
    edge_count: int
    nodes: List[GraphNode]
    edges: List[GraphEdge]
    dedup_clusters: Dict[str, List[str]] = {}
